library/core/src/char/methods.rs

```diff
@@ -1510,7 +1510,10 @@ impl char {
     #[rustc_const_stable(feature = "const_ascii_ctype_on_intrinsics", since = "1.47.0")]
     #[inline]
     pub const fn is_ascii_hexdigit(&self) -> bool {
-        matches!(*self, '0'..='9' | 'A'..='F' | 'a'..='f')
+        // Bitwise or converts A-Z to a-z, avoiding need for branches in compiled code.
+        const A: u32 = 'a' as u32;
+        const F: u32 = 'f' as u32;
+        matches!(*self, '0'..='9') || matches!(*self as u32 | 0x20, A..=F)
```

Comment:
Does this end up optimising the same if you refactor it to just call the `u8` method? Something like:

```rust
u8::try_from(*self).map_or(false, |c| c.is_ascii_hexdigit())
```

Reply:
No, because it has a branch for the "is it < 256?" part: <https://rust.godbolt.org/z/x13Yqx7Kx>

That said, something like that might actually perform really well, because it'd probably branch predict well.

For example, imagine something like this:
```rust
if *self <= 'f' {
    let c = u8::try_from(*this).unwrap();
    c.is_ascii_hexdigit()
} else {
    false
}
```

Yes there's a jump, but it probably predicts *great*.  Or at least as well as whatever jump you get from checking this condition in the first place...

Reply:
Ah, I was hoping that LLVM would be smart enough to optimise that out (since all other paths also involve a <N<256 comparison) but I guess not.

Reply:
I filed an LLVM bug about this: https://github.com/llvm/llvm-project/issues/60683

Even in the simpler case of `is_ascii_digit` it doesn't do as well as it ought to.